File: automaton/deploy/android/app/src/main/kotlin/com/automaton/client/AutomatonClient.kt
// AutomatonClient: OkHttp + kotlinx.serialization wrapper.
//
// Surface mirrors automaton/client.py and the Swift AutomatonClient so
// the Android app talks to the same HTTP API the CLI and iOS app do.
// All methods are suspend functions that run on the IO dispatcher.

package com.automaton.client

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import java.security.MessageDigest
import java.security.cert.X509Certificate
import javax.net.ssl.SSLContext
import javax.net.ssl.X509TrustManager

sealed class AutomatonError(msg: String) : Exception(msg) {
    class InvalidUrl(url: String) : AutomatonError("Invalid URL: $url")
    class Unauthorized : AutomatonError("Unauthorized — check the bearer token.")
    class NotFound(path: String) : AutomatonError("Not found: $path")
    class Server(status: Int, body: String) : AutomatonError("Server $status: $body")
    class Transport(cause: Throwable) : AutomatonError("Transport: ${cause.message}")
    class Decoding(cause: Throwable) : AutomatonError("Decoding: ${cause.message}")
}

private val JSON = Json {
    ignoreUnknownKeys = true
    coerceInputValues = true
}

class AutomatonClient(
    val baseUrl: String,
    private val token: String?,
    /** SHA-256 fingerprint of the leaf cert (hex). Set when using the
     *  self-signed cert from `automaton tls init` without installing
     *  it as a trusted CA on the device. */
    private val pinnedCertSHA256: String? = null,
) {
    private val http: OkHttpClient by lazy { buildHttp() }

    // ------------------------------------------------------------------ //
    // public methods                                                       //
    // ------------------------------------------------------------------ //

    suspend fun health(): Boolean = withContext(Dispatchers.IO) {
        val body = get<Map<String, Boolean>>("/healthz")
        body["ok"] ?: false
    }

    suspend fun runs(): List<RunSummary> = withContext(Dispatchers.IO) {
        get("/api/runs")
    }

    suspend fun runDetail(runId: Int): RunDetail = withContext(Dispatchers.IO) {
        get("/api/run/$runId")
    }

    /** Synthesize a workflow list from recent runs — server has no GET
     *  /api/workflows endpoint today (write-only). Mirrors the Swift client. */
    suspend fun workflows(): List<String> = withContext(Dispatchers.IO) {
        runs().map { it.workflow }.distinct().sorted()
    }

    suspend fun trigger(workflow: String,
                        payload: Map<String, Any>? = null): Int =
        withContext(Dispatchers.IO) {
            val body = buildMap<String, Any> {
                if (payload != null) put("payload", payload)
            }
            val res: TriggerResult = postJson("/api/trigger/$workflow", body)
            res.runId
        }

    suspend fun signal(runId: Int, name: String,
                       payload: Map<String, Any>? = null): Int =
        withContext(Dispatchers.IO) {
            val body = buildMap<String, Any> {
                if (payload != null) put("payload", payload)
            }
            val res: SignalResult = postJson("/api/signals/$runId/$name", body)
            res.signalId
        }

    suspend fun cancel(runId: Int, reason: String? = null): Boolean =
        withContext(Dispatchers.IO) {
            val body = buildMap<String, Any> {
                if (reason != null) put("reason", reason)
            }
            val res: CancelResult = postJson("/api/run/$runId/cancel", body)
            res.cancelled
        }

    // ------------------------------------------------------------------ //
    // internals                                                            //
    // ------------------------------------------------------------------ //

    private inline fun <reified T> get(path: String): T {
        val req = Request.Builder()
            .url(baseUrl.trimEnd('/') + path)
            .apply { token?.let { addHeader("Authorization", "Bearer $it") } }
            .build()
        return send(req)
    }

    private inline fun <reified T> postJson(path: String,
                                             body: Map<String, Any>): T {
        val jsonBody = JSON.encodeToString(
            kotlinx.serialization.serializer(), body
        ).toRequestBody("application/json".toMediaType())
        val req = Request.Builder()
            .url(baseUrl.trimEnd('/') + path)
            .post(jsonBody)
            .apply { token?.let { addHeader("Authorization", "Bearer $it") } }
            .build()
        return send(req)
    }

    private inline fun <reified T> send(req: Request): T {
        val resp: Response = try {
            http.newCall(req).execute()
        } catch (e: Exception) {
            throw AutomatonError.Transport(e)
        }
        resp.use { r ->
            val text = r.body?.string() ?: ""
            when (r.code) {
                in 200..299 -> Unit
                401 -> throw AutomatonError.Unauthorized()
                404 -> throw AutomatonError.NotFound(req.url.encodedPath)
                else -> throw AutomatonError.Server(r.code, text.take(240))
            }
            return try {
                JSON.decodeFromString(text)
            } catch (e: Exception) {
                throw AutomatonError.Decoding(e)
            }
        }
    }

    private fun buildHttp(): OkHttpClient {
        val builder = OkHttpClient.Builder()
        if (pinnedCertSHA256 != null) {
            val fp = pinnedCertSHA256.lowercase().replace(":", "")
            val tm = object : X509TrustManager {
                override fun checkClientTrusted(chain: Array<X509Certificate>, t: String) = Unit
                override fun getAcceptedIssuers(): Array<X509Certificate> = emptyArray()
                override fun checkServerTrusted(chain: Array<X509Certificate>, t: String) {
                    val observed = chain.firstOrNull()?.let {
                        MessageDigest.getInstance("SHA-256")
                            .digest(it.encoded)
                            .joinToString("") { b -> "%02x".format(b) }
                    } ?: throw Exception("empty cert chain")
                    if (observed != fp) throw Exception(
                        "cert fingerprint mismatch: expected $fp, got $observed"
                    )
                }
            }
            val ctx = SSLContext.getInstance("TLS").apply {
                init(null, arrayOf(tm), null)
            }
            builder
                .sslSocketFactory(ctx.socketFactory, tm)
                .hostnameVerifier { _, _ -> true }   // pinning by fingerprint; hostname check redundant
        }
        return builder.build()
    }
}

File: automaton/deploy/ios/Sources/AutomatonKit/AutomatonClient.swift
// AutomatonClient: async/await wrapper over URLSession.
//
// Surface mirrors automaton/client.py so the iOS app talks to the
// same HTTP API the CLI does. Methods all return Codable Sendable
// types so they cross actor boundaries cleanly.

import Foundation

public enum AutomatonError: Error, Sendable {
    case invalidURL
    case unauthorized
    case notFound(String)
    case server(status: Int, body: String)
    case transport(underlying: String)
    case decoding(underlying: String)
}

public struct AutomatonClient: Sendable {
    public let baseURL: URL
    public let token: String?
    /// SHA-256 fingerprint of the leaf cert, hex-encoded. Set this when
    /// the server uses the self-signed cert from `automaton tls init`
    /// and you don't want to install it as a trusted CA on the device.
    public let pinnedCertSHA256: String?

    public init(baseURL: URL, token: String?,
                pinnedCertSHA256: String? = nil) {
        self.baseURL = baseURL
        self.token = token
        self.pinnedCertSHA256 = pinnedCertSHA256
    }

    // MARK: - public methods

    public func health() async throws -> Bool {
        let body: [String: Bool] = try await get("/healthz")
        return body["ok"] ?? false
    }

    public func runs() async throws -> [RunSummary] {
        return try await get("/api/runs")
    }

    public func runDetail(_ runId: Int) async throws -> RunDetail {
        return try await get("/api/run/\(runId)")
    }

    public func workflows() async throws -> [String] {
        // The server doesn't have a /api/workflows GET today (write-only).
        // Synthesize the list from /api/runs' workflow column.
        let recents = try await runs()
        return Array(Set(recents.map(\.workflow))).sorted()
    }

    public func trigger(_ workflow: String,
                         payload: [String: Any]? = nil) async throws -> Int {
        var body: [String: Any] = [:]
        if let payload = payload { body["payload"] = payload }
        let res: TriggerResult = try await postJSON(
            "/api/trigger/\(workflow)", body: body
        )
        return res.runId
    }

    public func signal(_ runId: Int, name: String,
                        payload: [String: Any]? = nil) async throws -> Int {
        var body: [String: Any] = [:]
        if let payload = payload { body["payload"] = payload }
        let res: SignalResult = try await postJSON(
            "/api/signals/\(runId)/\(name)", body: body
        )
        return res.signalId
    }

    public func cancel(_ runId: Int, reason: String? = nil) async throws -> Bool {
        var body: [String: Any] = [:]
        if let reason = reason { body["reason"] = reason }
        let res: CancelResult = try await postJSON(
            "/api/run/\(runId)/cancel", body: body
        )
        return res.cancelled
    }

    // MARK: - internals

    private func get<T: Decodable>(_ path: String) async throws -> T {
        var req = URLRequest(url: baseURL.appendingPathComponent(path))
        if let token = token {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        return try await send(req)
    }

    private func postJSON<T: Decodable>(_ path: String,
                                          body: [String: Any]) async throws -> T {
        var req = URLRequest(url: baseURL.appendingPathComponent(path))
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let token = token {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        if !body.isEmpty {
            req.httpBody = try JSONSerialization.data(withJSONObject: body)
        }
        return try await send(req)
    }

    private func send<T: Decodable>(_ req: URLRequest) async throws -> T {
        let (data, resp): (Data, URLResponse)
        do {
            (data, resp) = try await session.data(for: req)
        } catch {
            throw AutomatonError.transport(underlying: String(describing: error))
        }
        guard let http = resp as? HTTPURLResponse else {
            throw AutomatonError.transport(underlying: "non-HTTP response")
        }
        switch http.statusCode {
        case 200..<300: break
        case 401:       throw AutomatonError.unauthorized
        case 404:       throw AutomatonError.notFound(req.url?.path ?? "")
        default:
            let text = String(data: data, encoding: .utf8) ?? ""
            throw AutomatonError.server(status: http.statusCode, body: text)
        }
        let decoder = JSONDecoder()
        decoder.keyDecodingStrategy = .convertFromSnakeCase
        do {
            return try decoder.decode(T.self, from: data)
        } catch {
            throw AutomatonError.decoding(underlying: String(describing: error))
        }
    }

    /// Lazy URLSession so the (optional) cert-pinning delegate is wired
    /// in exactly once per client instance.
    private var session: URLSession {
        if let fp = pinnedCertSHA256 {
            return URLSession(
                configuration: .default,
                delegate: PinnedCertDelegate(fingerprint: fp),
                delegateQueue: nil
            )
        }
        return .shared
    }
}

// MARK: - cert pinning

private final class PinnedCertDelegate: NSObject, URLSessionDelegate, Sendable {
    let fingerprint: String
    init(fingerprint: String) { self.fingerprint = fingerprint.lowercased() }

    func urlSession(_ session: URLSession,
                    didReceive challenge: URLAuthenticationChallenge,
                    completionHandler: @escaping (URLSession.AuthChallengeDisposition,
                                                   URLCredential?) -> Void) {
        guard challenge.protectionSpace.authenticationMethod
              == NSURLAuthenticationMethodServerTrust,
              let trust = challenge.protectionSpace.serverTrust,
              let cert = SecTrustGetCertificateAtIndex(trust, 0) else {
            completionHandler(.cancelAuthenticationChallenge, nil)
            return
        }
        let data = SecCertificateCopyData(cert) as Data
        let observed = data.sha256HexString()
        if observed == fingerprint {
            completionHandler(.useCredential, URLCredential(trust: trust))
        } else {
            completionHandler(.cancelAuthenticationChallenge, nil)
        }
    }
}

import CryptoKit

private extension Data {
    func sha256HexString() -> String {
        let digest = SHA256.hash(data: self)
        return digest.map { String(format: "%02x", $0) }.joined()
    }
}
